### DISCUSSION TITLE
Allow interfaces to be used as module configs

### DISCUSSION THREAD

**Discussion by marcus-sa** [marcj/deepkit#479] [Ideas]
Use case: Creating Deepkit modules for 3rd party packages.
```ts
import { ApolloServerOptionsWithSchema } from '@apollo/server';

interface ApolloGraphQLConfig
  extends
    Pick<
      ApolloServerOptionsWithSchema<{}>,
      | 'introspection'
      | 'nodeEnv'
      | 'allowBatchedHttpRequests'
      | 'csrfPrevention'
      | 'includeStacktraceInErrorResponses'
      | 'apollo'
      | 'stopOnTerminationSignals'
      | 'persistedQueries'
    > {}

class ApolloGraphQLModule extends createModule({
  config: typeOf<ApolloGraphQLConfig>(),
}) {}
```

**Comment by marcj** [comment]
that's an interesting one that will surely come up more and more. the problem is that we do not have type information at runtime from `@apollo/server`. A workaround we do in types from d.ts files (from global types like Pick, Omit, etc) is that we simply inline them at the top of the file. We could do the same for apollo, but inlining doesn't support everything, like function/class/symbol references won't be there. we would have to manually add imports from apollo/server to make them available which won't work always (what if they are not exported?). so, if the imported type is one without references to classes/functions/symbols we can make it work, or maybe we find a better solution to inlining.

**Reply by marcus-sa** [reply]
@marcj if we redid the whole configuration of the type compiler and added a `deepkitTypeCompilerOptions` section to `tsconfig.json` (similarly to `angularCompilerOptions`), then we could add a new option called `external`. This would accept an input with the exported types to include from a third party import, so that they could be inlined (similarly to how were doing with global types).

Suggestion one
```json
{
  "deepkitTypeCompilerOptions": {
    "reflection": true,
    "external": [
      "@apollo/server", // for all imported types referenced
      ["@apollo/server", ["ApolloServerOptionsWithSchema"]] // for specific imported types referenced 
    ]
  }
}
```
Suggestion two
```json
{
  "deepkitTypeCompilerOptions": {
    "reflection": true,
    "external": {
      "@apollo/server": "*", // for all imported types referenced
      "@apollo/server": ["ApolloServerOptionsWithSchema"] // for specific imported types referenced 
    }
  }
}
```
